File: analysis/r/sales_analysis.R
# =====================================================================
# Sales Performance Analysis  —  R
# ---------------------------------------------------------------------
# Reproduces the enriched fact table and every KPI in the three
# dashboards using the tidyverse.
#
# Assumptions:
#   * Revenue is on NET units (sold - returned).
#   * Each Sales Person ID is treated as its Store ID (1:1 here);
#     store revenue is attributed via the sales person's store.
#   * Customer age is computed as of 2023-12-31.
#
# Usage:  Rscript sales_analysis.R /path/to/csv/folder
# =====================================================================

suppressPackageStartupMessages({
  library(dplyr)
  library(readr)
  library(lubridate)
  library(tidyr)
})

args   <- commandArgs(trailingOnly = TRUE)
folder <- if (length(args) >= 1) args[1] else "."
fp     <- function(name) file.path(folder, name)

# ---------------------------------------------------------------------
# 1. LOAD
# ---------------------------------------------------------------------
fact <- read_csv(fp("fact_table.csv"),           show_col_types = FALSE)
cust <- read_csv(fp("customers_table.csv"),      show_col_types = FALSE)
prod <- read_csv(fp("products_table.csv"),       show_col_types = FALSE)
sp   <- read_csv(fp("sales_persons_table.csv"),  show_col_types = FALSE)
tgt  <- read_csv(fp("monthly_store_targets.csv"),show_col_types = FALSE)

fact <- fact %>% mutate(`Order Date` = mdy(`Order Date`))
cust <- cust %>% mutate(`Date of Birth` = mdy(`Date of Birth`)) %>%
  rename(`Cust DOB` = `Date of Birth`)
sp   <- sp %>% select(-`Date of Birth`)
tgt  <- tgt %>% mutate(Month = mdy(Month))

# ---------------------------------------------------------------------
# 2. ENRICHED FACT TABLE
# ---------------------------------------------------------------------
ref_date <- as.Date("2023-12-31")

df <- fact %>%
  left_join(prod, by = "Product ID") %>%
  left_join(cust, by = "Customer ID") %>%
  left_join(sp,   by = "Sales Person ID") %>%
  mutate(
    `Net Quantity` = `Quantity Sold` - `Quantity Returned`,
    Revenue        = `Net Quantity` * `Sales Price`,
    COGS           = `Net Quantity` * `Cost Price`,
    Profit         = Revenue - COGS,
    Age            = as.integer(floor(as.numeric(ref_date - `Cust DOB`) / 365.25)),
    `Age Group`    = cut(Age,
                         breaks = c(0, 25, 35, 45, 55, 65, Inf),
                         labels = c("Under 25","25-34","35-44","45-54","55-64","65+"),
                         right  = FALSE),
    MonthStart = floor_date(`Order Date`, "month"),
    MonthName  = month(`Order Date`, label = TRUE, abbr = TRUE),
    MonthNum   = month(`Order Date`),
    Weekday    = wday(`Order Date`, label = TRUE, abbr = FALSE, week_start = 1),
    Quarter    = paste0("Q", quarter(`Order Date`)),
    IsWeekend  = wday(`Order Date`, week_start = 1) >= 6,
    `Store Name` = coalesce(`Store Name`, "Unknown")
  )

# ---------------------------------------------------------------------
# 3. HEADLINE KPIs
# ---------------------------------------------------------------------
kpis <- df %>% summarise(
  `Total Revenue`      = sum(Revenue),
  `Total Profit`       = sum(Profit),
  `Profit Margin`      = sum(Profit) / sum(Revenue),
  `Total Units Sold`   = sum(`Quantity Sold`),
  `Total Returned`     = sum(`Quantity Returned`),
  `Overall Return Rate`= sum(`Quantity Returned`) / sum(`Quantity Sold`)
)

# ---------------------------------------------------------------------
# 4. DASHBOARD 1 — CUSTOMER & PRODUCT
# ---------------------------------------------------------------------
d1_gender <- df %>%
  group_by(Gender) %>%
  summarise(Profit = sum(Profit), Revenue = sum(Revenue), .groups = "drop")

d1_age <- df %>%
  group_by(`Age Group`) %>%
  summarise(Profit = sum(Profit), Revenue = sum(Revenue),
            Orders = n(), .groups = "drop") %>%
  mutate(`Avg Spend` = Revenue / Orders)

d1_time <- df %>%
  group_by(MonthStart) %>%
  summarise(Profit = sum(Profit), Revenue = sum(Revenue), .groups = "drop") %>%
  arrange(MonthStart) %>%
  mutate(`MoM Growth` = (Profit - lag(Profit)) / lag(Profit),
         `MoM Growth` = replace_na(`MoM Growth`, 0))

d1_weekday <- df %>%
  group_by(Weekday) %>%
  summarise(Profit = sum(Profit), .groups = "drop") %>%
  arrange(Weekday)

product_tbl <- df %>%
  group_by(`Product Name`) %>%
  summarise(Profit      = sum(Profit),
            Revenue     = sum(Revenue),
            QtySold     = sum(`Quantity Sold`),
            QtyReturned = sum(`Quantity Returned`), .groups = "drop") %>%
  mutate(`Return Rate` = QtyReturned / QtySold)

d1_top_profit  <- product_tbl %>% arrange(desc(Profit))        %>% slice_head(n = 10)
d1_top_selling <- product_tbl %>% arrange(desc(QtySold))       %>% slice_head(n = 10)
d1_top_returns <- product_tbl %>% arrange(desc(`Return Rate`)) %>% slice_head(n = 10)

# ---------------------------------------------------------------------
# 5. DASHBOARD 2 — STORE BUDGET VS REVENUE
# ---------------------------------------------------------------------
sp2 <- sp %>% mutate(`Store ID` = `Sales Person ID`)

store_rev <- df %>%
  group_by(`Store Name`) %>%
  summarise(Revenue = sum(Revenue), .groups = "drop")

store_tgt <- tgt %>%
  group_by(`Store ID`) %>%
  summarise(Target = sum(`Monthly Target`), .groups = "drop") %>%
  left_join(sp2 %>% select(`Store ID`, `Store Name`), by = "Store ID")

d2_store_bvr <- store_rev %>%
  left_join(store_tgt %>% select(`Store Name`, Target), by = "Store Name") %>%
  mutate(Variance = Revenue - Target,
         `% to Target` = Revenue / Target) %>%
  arrange(desc(Revenue))

month_rev <- df  %>% group_by(MonthStart) %>%
  summarise(Revenue = sum(Revenue), .groups = "drop")
month_tgt <- tgt %>% mutate(MonthStart = floor_date(Month, "month")) %>%
  group_by(MonthStart) %>%
  summarise(Target = sum(`Monthly Target`), .groups = "drop")

d2_month <- full_join(month_rev, month_tgt, by = "MonthStart") %>%
  arrange(MonthStart) %>%
  mutate(across(c(Revenue, Target), ~replace_na(., 0)),
         Variance = Revenue - Target)

# ---------------------------------------------------------------------
# 6. DASHBOARD 3 — REVENUE ANALYSIS
# ---------------------------------------------------------------------
d3_quarter <- df %>%
  group_by(Quarter) %>%
  summarise(Revenue = sum(Revenue), .groups = "drop") %>%
  mutate(`Avg Revenue` = mean(Revenue),
         `Variance vs Avg` = Revenue - mean(Revenue))

d3_weekend <- df %>%
  mutate(Type = if_else(IsWeekend, "Weekend", "Weekday")) %>%
  group_by(Type) %>%
  summarise(Revenue = sum(Revenue), .groups = "drop")

# d3 monthly revenue vs target == d2_month (reused)

# ---------------------------------------------------------------------
# 7. OUTPUT
# ---------------------------------------------------------------------
cat("=== HEADLINE KPIs ===\n");           print(as.data.frame(kpis))
cat("\n=== Profit by Gender ===\n");       print(as.data.frame(d1_gender))
cat("\n=== Avg Spend by Age Group ===\n"); print(as.data.frame(d1_age))
cat("\n=== Store Revenue vs Target ===\n");print(as.data.frame(d2_store_bvr))
cat("\n=== Quarterly Revenue ===\n");      print(as.data.frame(d3_quarter))
cat("\n=== Weekday vs Weekend ===\n");     print(as.data.frame(d3_weekend))

# Optional: write each table to CSV
# write_csv(d2_store_bvr, "out_store_bvr.csv")  # etc.

File: analysis/python/sales_analysis.py
"""
Sales Performance Analysis — Python Pipeline
=============================================
Builds the enriched fact table and every KPI used in the three dashboards.

Inputs (CSV):
  fact_table.csv          - one row per order line
  customers_table.csv     - Customer ID, name, Gender, Location, Date of Birth
  products_table.csv      - Product ID, name, Category, Sales Price, Cost Price
  sales_persons_table.csv - Sales Person ID, name, Store Name  (Sales Person ID == Store ID)
  monthly_store_targets.csv - Store ID, Month, Monthly Target

Run:  python sales_analysis.py /path/to/csv/folder
"""

import sys
import pandas as pd
import numpy as np

# ----------------------------------------------------------------------
# 1. LOAD
# ----------------------------------------------------------------------
def load_data(folder):
    f = lambda name: pd.read_csv(f"{folder.rstrip('/')}/{name}")
    fact = f("fact_table.csv")
    cust = f("customers_table.csv")
    prod = f("products_table.csv")
    sp   = f("sales_persons_table.csv")
    tgt  = f("monthly_store_targets.csv")

    fact["Order Date"]    = pd.to_datetime(fact["Order Date"])
    cust["Date of Birth"] = pd.to_datetime(cust["Date of Birth"])
    tgt["Month"]          = pd.to_datetime(tgt["Month"])

    # Both customers and sales persons carry "Date of Birth" -> avoid a merge collision
    cust = cust.rename(columns={"Date of Birth": "Cust DOB"})
    sp   = sp.drop(columns=["Date of Birth"])
    return fact, cust, prod, sp, tgt


# ----------------------------------------------------------------------
# 2. BUILD ENRICHED FACT TABLE
# ----------------------------------------------------------------------
def build_fact(fact, cust, prod, sp):
    df = (fact
          .merge(prod, on="Product ID",      how="left")
          .merge(cust, on="Customer ID",     how="left")
          .merge(sp,   on="Sales Person ID", how="left"))

    # Core money metrics. Revenue is on NET units (sold minus returned),
    # so returns are already reflected in revenue and profit.
    df["Net Quantity"]  = df["Quantity Sold"] - df["Quantity Returned"]
    df["Revenue"]       = df["Net Quantity"] * df["Sales Price"]
    df["COGS"]          = df["Net Quantity"] * df["Cost Price"]
    df["Profit"]        = df["Revenue"] - df["COGS"]
    df["Gross Revenue"] = df["Quantity Sold"] * df["Sales Price"]

    # Customer age as of end of the analysis year
    ref = pd.Timestamp("2023-12-31")
    df["Age"] = ((ref - df["Cust DOB"]).dt.days / 365.25).astype(int)
    bins   = [0, 25, 35, 45, 55, 65, 200]
    labels = ["Under 25", "25-34", "35-44", "45-54", "55-64", "65+"]
    df["Age Group"] = pd.cut(df["Age"], bins=bins, labels=labels, right=False)

    # Date dimensions
    df["Month"]      = df["Order Date"].dt.to_period("M").dt.to_timestamp()
    df["MonthName"]  = df["Order Date"].dt.strftime("%b")
    df["MonthNum"]   = df["Order Date"].dt.month
    df["Weekday"]    = df["Order Date"].dt.day_name()
    df["WeekdayNum"] = df["Order Date"].dt.weekday          # Mon=0 .. Sun=6
    df["Quarter"]    = "Q" + df["Order Date"].dt.quarter.astype(str)
    df["IsWeekend"]  = df["WeekdayNum"].isin([5, 6])
    df["Store Name"] = df["Store Name"].fillna("Unknown")
    return df


# ----------------------------------------------------------------------
# 3. DASHBOARD 1 — CUSTOMER & PRODUCT
# ----------------------------------------------------------------------
def customer_and_product(df):
    out = {}

    # Profit & revenue by gender
    out["gender"] = (df.groupby("Gender", observed=True)
                       .agg(Profit=("Profit", "sum"), Revenue=("Revenue", "sum"))
                       .reset_index())

    # Profit, revenue & average spend by age group
    age = (df.groupby("Age Group", observed=True)
             .agg(Profit=("Profit", "sum"),
                  Revenue=("Revenue", "sum"),
                  Orders=("Order Date", "count"))
             .reset_index())
    age["Avg Spend"] = age["Revenue"] / age["Orders"]
    out["age_group"] = age

    # Profitability over time + month-over-month growth
    ptime = (df.groupby("Month")
               .agg(Profit=("Profit", "sum"), Revenue=("Revenue", "sum"))
               .reset_index()
               .sort_values("Month"))
    ptime["MoM Growth"] = ptime["Profit"].pct_change().fillna(0)
    out["profit_over_time"] = ptime

    # Profit by weekday
    out["weekday"] = (df.groupby(["WeekdayNum", "Weekday"])
                        .agg(Profit=("Profit", "sum"))
                        .reset_index()
                        .sort_values("WeekdayNum"))

    # Product-level table -> top sellers, most profitable, highest return rate
    p = (df.groupby("Product Name")
           .agg(Profit=("Profit", "sum"),
                Revenue=("Revenue", "sum"),
                QtySold=("Quantity Sold", "sum"),
                QtyReturned=("Quantity Returned", "sum"))
           .reset_index())
    p["Return Rate"] = p["QtyReturned"] / p["QtySold"]
    out["top_profit"]  = p.sort_values("Profit",      ascending=False).head(10)
    out["top_selling"] = p.sort_values("QtySold",     ascending=False).head(10)
    out["top_returns"] = p.sort_values("Return Rate", ascending=False).head(10)
    return out


# ----------------------------------------------------------------------
# 4. DASHBOARD 2 — STORE BUDGET VS REVENUE
# ----------------------------------------------------------------------
def store_budget(df, sp, tgt):
    out = {}
    sp = sp.copy()
    sp["Store ID"] = sp["Sales Person ID"]   # 1:1 mapping in this dataset

    store_rev = (df.groupby("Store Name")
                   .agg(Revenue=("Revenue", "sum"))
                   .reset_index())
    store_tgt = (tgt.groupby("Store ID")
                    .agg(Target=("Monthly Target", "sum"))
                    .reset_index()
                    .merge(sp[["Store ID", "Store Name"]], on="Store ID"))

    bvr = store_rev.merge(store_tgt[["Store Name", "Target"]], on="Store Name", how="left")
    bvr["Variance"]     = bvr["Revenue"] - bvr["Target"]
    bvr["% to Target"]  = bvr["Revenue"] / bvr["Target"]
    out["store_bvr"] = bvr.sort_values("Revenue", ascending=False)

    # Month-by-month total revenue vs target
    month_rev = df.groupby("Month").agg(Revenue=("Revenue", "sum")).reset_index()
    month_tgt = tgt.groupby("Month").agg(Target=("Monthly Target", "sum")).reset_index()
    mbm = (month_rev.merge(month_tgt, on="Month", how="outer")
                    .sort_values("Month").fillna(0))
    mbm["Variance"] = mbm["Revenue"] - mbm["Target"]
    out["month_by_month"] = mbm
    return out


# ----------------------------------------------------------------------
# 5. DASHBOARD 3 — REVENUE ANALYSIS
# ----------------------------------------------------------------------
def revenue_analysis(df):
    out = {}

    # Quarterly revenue vs the quarterly average
    q = df.groupby("Quarter").agg(Revenue=("Revenue", "sum")).reset_index()
    q["Avg Revenue"] = q["Revenue"].mean()
    out["quarterly"] = q

    # Weekday vs weekend revenue
    ww = (df.groupby("IsWeekend")
            .agg(Revenue=("Revenue", "sum"))
            .reset_index())
    ww["Type"] = ww["IsWeekend"].map({True: "Weekend", False: "Weekday"})
    out["weekday_weekend"] = ww[["Type", "Revenue"]]

    # Monthly revenue vs target reuses the Dashboard-2 month-by-month table
    return out


# ----------------------------------------------------------------------
# 6. HEADLINE KPIs
# ----------------------------------------------------------------------
def headline_kpis(df):
    return {
        "Total Revenue":      df["Revenue"].sum(),
        "Total Profit":       df["Profit"].sum(),
        "Profit Margin":      df["Profit"].sum() / df["Revenue"].sum(),
        "Total Units Sold":   df["Quantity Sold"].sum(),
        "Total Returned":     df["Quantity Returned"].sum(),
        "Overall Return Rate": df["Quantity Returned"].sum() / df["Quantity Sold"].sum(),
    }


# ----------------------------------------------------------------------
# MAIN
# ----------------------------------------------------------------------
def main(folder):
    fact, cust, prod, sp, tgt = load_data(folder)
    df = build_fact(fact, cust, prod, sp)

    kpis = headline_kpis(df)
    d1   = customer_and_product(df)
    d2   = store_budget(df, sp, tgt)
    d3   = revenue_analysis(df)

    print("\n=== HEADLINE KPIs ===")
    for k, v in kpis.items():
        print(f"  {k:<20} {v:,.2f}")

    print("\n=== Profit by Gender ===");        print(d1["gender"].to_string(index=False))
    print("\n=== Avg Spend by Age Group ===");  print(d1["age_group"].to_string(index=False))
    print("\n=== Store Revenue vs Target ===");  print(d2["store_bvr"].to_string(index=False))
    print("\n=== Quarterly Revenue ===");        print(d3["quarterly"].to_string(index=False))
    print("\n=== Weekday vs Weekend ===");       print(d3["weekday_weekend"].to_string(index=False))

    return df, kpis, d1, d2, d3


if __name__ == "__main__":
    folder = sys.argv[1] if len(sys.argv) > 1 else "."
    main(folder)
